azurerm/resource_arm_private_dns_srv_record.go

```diff
@@ -0,0 +1,262 @@
+		Name: &name,
+		RecordSetProperties: &privatedns.RecordSetProperties{
+			Metadata:   tags.Expand(t),
+			TTL:        &ttl,
+			SrvRecords: expandAzureRmPrivateDnsSrvRecords(d),
+		},
+	}
+
+	eTag := ""
+	ifNoneMatch := "" // set to empty to allow updates to records after creation
+	if _, err := client.CreateOrUpdate(ctx, resGroup, zoneName, privatedns.SRV, name, parameters, eTag, ifNoneMatch); err != nil {
+		return fmt.Errorf("Error creating/updating Private DNS SRV Record %q (Zone %q / Resource Group %q): %s", name, zoneName, resGroup, err)
+	}
+
+	resp, err := client.Get(ctx, resGroup, zoneName, privatedns.SRV, name)
+	if err != nil {
+		return fmt.Errorf("Error retrieving Private DNS SRV Record %q (Zone %q / Resource Group %q): %s", name, zoneName, resGroup, err)
+	}
+
+	if resp.ID == nil {
+		return fmt.Errorf("Cannot read Private DNS SRV Record %s (resource group %s) ID", name, resGroup)
+	}
+
+	d.SetId(*resp.ID)
+
+	return resourceArmPrivateDnsSrvRecordRead(d, meta)
+}
+
+func resourceArmPrivateDnsSrvRecordRead(d *schema.ResourceData, meta interface{}) error {
+	dnsClient := meta.(*ArmClient).PrivateDns.RecordSetsClient
+	ctx, cancel := timeouts.ForRead(meta.(*ArmClient).StopContext, d)
+	defer cancel()
+
+	id, err := azure.ParseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	resGroup := id.ResourceGroup
+	name := id.Path["SRV"]
+	zoneName := id.Path["privateDnsZones"]
+
+	resp, err := dnsClient.Get(ctx, resGroup, zoneName, privatedns.SRV, name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error reading Private DNS SRV record %s: %+v", name, err)
+	}
+
+	d.Set("name", name)
+	d.Set("resource_group_name", resGroup)
+	d.Set("zone_name", zoneName)
+	d.Set("ttl", resp.TTL)
+
+	if err := d.Set("record", flattenAzureRmPrivateDnsSrvRecords(resp.SrvRecords)); err != nil {
+		return err
+	}
+
+	return tags.FlattenAndSet(d, resp.Metadata)
+}
+
+func resourceArmPrivateDnsSrvRecordDelete(d *schema.ResourceData, meta interface{}) error {
+	dnsClient := meta.(*ArmClient).PrivateDns.RecordSetsClient
+	ctx, cancel := timeouts.ForDelete(meta.(*ArmClient).StopContext, d)
+	defer cancel()
+
+	id, err := azure.ParseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	resGroup := id.ResourceGroup
+	name := id.Path["SRV"]
+	zoneName := id.Path["privateDnsZones"]
+
+	_, err = dnsClient.Delete(ctx, resGroup, zoneName, privatedns.SRV, name, "")
+	if err != nil {
+		return fmt.Errorf("Error deleting Private DNS SRV Record %s: %+v", name, err)
+	}
+
+	return nil
+}
+
+func flattenAzureRmPrivateDnsSrvRecords(records *[]privatedns.SrvRecord) []map[string]interface{} {
+	results := make([]map[string]interface{}, 0)
+
+	if records != nil {
+		for _, record := range *records {
+			results = append(results, map[string]interface{}{
+				"priority": *record.Priority,
+				"weight":   *record.Weight,
+				"port":     *record.Port,
+				"target":   *record.Target,
```

Comment:
we will need to nil check these de-references

Reply:
Do we really? The incoming record should always be complete as the content is validated in the Schema

Reply:
the Azure API returns the representation of a resource at the last time it was modified, rather then the latest representation for the API version; as such we've seen (multiple) instances where the API response differs from the schema defined in the SDK, which means these values will be nil - which is why we're cautious parsing the response back from the API - so unfortunately yes (thanks for pushing this)